### PR TITLE
Split AppWrapper controller into separate dispatcher and runner

### PR DESCRIPTION
**Pull request by dgrove-oss**
Progress towards multi-cluster MCAD by splitting the AppWrapper reconciler into two separate reconcilers that handle disjoint state transitions in the lifecycle of the AppWrapper.

### CODE REVIEW

**Comment by dgrove-oss** [comment]
Ready for review.

Made a pass this morning to update comments and to add additional steps to the Running state to allow a place for the synch/unsynch operations to take place in a true multi-cluster deployment.

File: api/v1beta1/appwrapper_types.go [hunk 1]
```diff
@@ -148,15 +148,21 @@ const (
...
 	Creating AppWrapperStep = "creating"
 
 	// The wrapped resources have been deployed successfully
 	Created AppWrapperStep = "created"
 
-	// MCAD is in the process of deleting the wrapped resources
+	// The MCAD runner is in the process of deleting the wrapped resources
 	Deleting AppWrapperStep = "deleting"
 
+	//The MCAD runner has returned control of an appwrapper to the dispatcher
```
Comment (tardieu) [inline api/v1beta1/appwrapper_types.go:163]: Missing a space between `//` and `The`

File: internal/controller/runner.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+	mcadv1beta1 "github.com/project-codeflare/mcad/api/v1beta1"
+)
+
+// AppWrapperReconciler responsible for the portion of the lifecycle that happens on the execution cluster
+type Runner struct {
+	AppWrapperReconciler
+}
+
+// permission to edit appwrappers
```
Comment (tardieu) [inline internal/controller/runner.go:43]: Some rbacs are duplicated between the dispatcher and runner. We should probably regroup the common rules into one stanza attached to the super type.

File: internal/controller/runner.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+	// append appWrapper ID to logger
+	ctx = withAppWrapper(ctx, appWrapper)
+
+	// abort and requeue reconciliation if reconciler cache is stale
+	if r.isStale(ctx, appWrapper) {
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// handle deletion
+	if !appWrapper.DeletionTimestamp.IsZero() {
```
Comment (tardieu) [inline internal/controller/runner.go:83]: Maybe also check that the runner finalizer is still there?

File: internal/controller/dispatcher.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+		}
+
+	case mcadv1beta1.Succeeded:
+		switch appWrapper.Status.Step {
+		case mcadv1beta1.Idle:
+			r.triggerDispatch()
+			return ctrl.Result{}, nil
+		}
+
+	case mcadv1beta1.Failed:
```
Comment (tardieu) [inline internal/controller/dispatcher.go:193]: We should transition `Failed, deleted` to `Failed, idle` by removing the placement object in multi-cluster mode.

**Review by tardieu — CHANGES_REQUESTED** [review]
One issue: the lack of a transition from `Failed, deleted` to `Failed, idle` plus a bunch of minor tweaks.

File: internal/controller/dispatcher.go (hunk 4 above)
Comment (dgrove-oss) [reply internal/controller/dispatcher.go:187]: done

File: internal/controller/runner.go (hunk 3 above)
Comment (dgrove-oss) [reply internal/controller/runner.go:77]: done

File: internal/controller/runner.go (hunk 2 above)
Comment (dgrove-oss) [reply internal/controller/runner.go]: done

**Review by tardieu — APPROVED** [review]
lgtm

**Pull request merged**